internal/snakegame/snakegame.go

```diff
@@ -0,0 +1,221 @@
+		p.moving(turn, exit)
+		p.foodInteraction()
+		p.clean()
+	}
+
+}
+
+func (p *SnakeGame) foodGenerator() {
+	x := rand.Intn(len(p.board))
+	y := rand.Intn(len(p.board[0]))
+
+	if p.board[x][y] == kEmptyCell {
+		p.food = point{x, y}
+	} else {
+		p.foodGenerator()
+	}
+}
+
+// Add snake and food to board matrix
+func (p *SnakeGame) updateBoard() {
+	for index := range p.snake.body {
+		if index == kHeadIndex {
+			p.board[p.snake.body[index].x][p.snake.body[index].y] = kHead
+		} else {
+			p.board[p.snake.body[index].x][p.snake.body[index].y] = kBody
+		}
+	}
+	p.board[p.food.x][p.food.y] = kApple
+}
+
+func (p *SnakeGame) showBoard() {
+	for i := range p.board {
+		for j := range p.board[i] {
+			fmt.Printf("%c", p.board[i][j])
+		}
+		fmt.Println()
+	}
+}
+
+func (p *SnakeGame) clean() {
+	for i := range p.board {
+		for j := range p.board[i] {
+			p.board[i][j] = kEmptyCell
+		}
+	}
+	cmd := exec.Command("clear")
+	cmd.Stdout = os.Stdout
+	cmd.Run()
+}
+
+func (p *SnakeGame) foodInteraction() {
+	if p.snake.isFoodEaten {
+		p.score += 1
+		p.snake.body = append(p.snake.body, p.snake.tail)
+		p.snake.isFoodEaten = false
+	}
+	if p.board[p.snake.body[kHeadIndex].x][p.snake.body[kHeadIndex].y] == kApple {
+		p.snake.isFoodEaten = true
+		p.pauseTime += 50
+		p.foodGenerator()
+	}
+	p.snake.tail = p.snake.body[len(p.snake.body)-1]
+}
+
+func (p *SnakeGame) wallInteraction() {
+	for i := range p.snake.body {
+		if p.snake.body[i].x > len(p.board)-1 {
+			p.snake.body[i].x -= len(p.board)
+		}
+		if p.snake.body[i].y > len(p.board[0])-1 {
+			p.snake.body[i].y -= len(p.board[0])
+		}
+		if p.snake.body[i].x < 0 {
+			p.snake.body[i].x += len(p.board)
+		}
+		if p.snake.body[i].y < 0 {
+			p.snake.body[i].y += len(p.board[0])
+		}
+	}
+}
+
+func (p *SnakeGame) crash(exit chan<- struct{}) {
+	for i := range p.snake.body {
+		if (p.snake.body[kHeadIndex].x == p.snake.body[i].x && kHeadIndex != i) && (p.snake.body[kHeadIndex].y == p.snake.body[i].y && kHeadIndex != i) {
+			exit <- struct{}{}
+		}
+	}
+
+}
+
+type point struct {
+	x int
+	y int
+}
+
+func (p *SnakeGame) UserControl(turn chan<- int, exit chan<- struct{}) {
+
+	keyData, err := keyboard.GetKeys(10)
+
+	if err != nil {
+		panic(err)
+	}
+	defer func() {
+		_ = keyboard.Close()
+	}()
+
+	for {
+		event := <-keyData
+		if event.Err != nil {
+			panic(event.Err)
+		}
+
+		switch event.Key {
+		case keyboard.KeyArrowUp:
+			if p.snake.currentDirectional == kRight || p.snake.currentDirectional == kLeft {
+				turn <- kUp
+			}
+		case keyboard.KeyArrowRight:
+			if p.snake.currentDirectional == kUp || p.snake.currentDirectional == kDown {
+				turn <- kRight
+			}
+		case keyboard.KeyArrowDown:
+			if p.snake.currentDirectional == kRight || p.snake.currentDirectional == kLeft {
+				turn <- kDown
+			}
+		case keyboard.KeyArrowLeft:
+			if p.snake.currentDirectional == kUp || p.snake.currentDirectional == kDown {
+				turn <- kLeft
+			}
+		case keyboard.KeyEsc:
+			exit <- struct{}{}
```

Comment:
As I see, you put something to "exit" channel from two different goroutines.
What if both of them will put to channel simultaneously?

Reply:
Refactored